pkg/vulnsrc/openeuler/openeuler.go

```diff
@@ -0,0 +1,267 @@
+			return xerrors.Errorf("failed to decode openEuler CVRF JSON: %w %+v", err, cvrf)
+		}
+		cvrfs = append(cvrfs, cvrf)
+		return nil
+	})
+	if err != nil {
+		return xerrors.Errorf("error in openEuler CVRF walk: %w", err)
+	}
+
+	if err = vs.save(cvrfs); err != nil {
+		return xerrors.Errorf("error in openEuler CVRF save: %w", err)
+	}
+
+	return nil
+}
+
+func (vs VulnSrc) save(cvrfs []Cvrf) error {
+	err := vs.dbc.BatchUpdate(func(tx *bolt.Tx) error {
+		return vs.commit(tx, cvrfs)
+	})
+	if err != nil {
+		return xerrors.Errorf("error in batch update: %w", err)
+	}
+	return nil
+}
+
+func (vs VulnSrc) commit(tx *bolt.Tx, cvrfs []Cvrf) error {
+	var uniqOSVers = make(map[string]struct{})
+	for _, cvrf := range cvrfs {
+		affectedPkgs := getAffectedPackages(cvrf.ProductTree)
+		if len(affectedPkgs) == 0 {
+			continue
+		}
+
+		for _, pkg := range affectedPkgs {
+			arches := lo.Uniq(pkg.Archs)
+			sort.Strings(arches)
+			advisory := types.Advisory{
+				FixedVersion: pkg.FixedVersion,
+				Arches:       arches,
+			}
+			// Don't put the same data source multiple times.
+			if _, ok := uniqOSVers[pkg.OSVer]; !ok {
+				uniqOSVers[pkg.OSVer] = struct{}{}
+				if err := vs.dbc.PutDataSource(tx, pkg.OSVer, source); err != nil {
+					return xerrors.Errorf("failed to put data source: %w", err)
+				}
+			}
+
+			if err := vs.dbc.PutAdvisoryDetail(tx, cvrf.Tracking.ID, pkg.Name,
+				[]string{pkg.OSVer}, advisory); err != nil {
+				return xerrors.Errorf("unable to save %s CVRF: %w", pkg.OSVer, err)
+			}
+		}
+
+		var references []string
+		for _, ref := range cvrf.References {
+			references = append(references, ref.URL)
+		}
+
+		severity := types.SeverityUnknown
+		for _, cvuln := range cvrf.Vulnerabilities {
+			for _, threat := range cvuln.Threats {
+				sev := severityFromThreat(threat.Severity)
+				if severity < sev {
+					severity = sev
+				}
+			}
+		}
+
+		vuln := types.VulnerabilityDetail{
+			References:  references,
+			Title:       cvrf.Title,
+			Description: getDetail(cvrf.Notes),
+			Severity:    severity,
+		}
+		if err := vs.dbc.PutVulnerabilityDetail(tx, cvrf.Tracking.ID, source.ID, vuln); err != nil {
+			return xerrors.Errorf("failed to save openEuler CVRF vulnerability: %w", err)
+		}
+
+		// for optimization
+		if err := vs.dbc.PutVulnerabilityID(tx, cvrf.Tracking.ID); err != nil {
+			return xerrors.Errorf("failed to save the vulnerability ID: %w", err)
+		}
+	}
+	return nil
+}
+
+func getAffectedPackages(productTree ProductTree) []Package {
+	var pkgs = make(map[string]Package) // pkgID => Package
+	for _, branch := range productTree.Branches {
+		// `src` pkgs are not installed in openEuler.
+		if branch.Type != "Package Arch" || branch.Name == "src" || branch.Name == "" {
+			continue
+		}
+		for _, production := range branch.Productions {
+			osVer := getOSVersion(production.CPE)
+			if osVer == "" {
+				log.Printf("unable to parse OS version: %s", production.CPE)
```

Comment:
nit: Log messages should be capitalized.

```suggestion
				log.Printf("Unable to parse OS version: %s", production.CPE)
```

Reply:
done